service-clients/src/main/java/com/google/cloud/pso/rag/vector/AlloyDB.java

```diff
@@ -0,0 +1,262 @@
+      case SearchRequest searchRequest -> searchSQLStatement(searchRequest);
+      case UpsertRequest upsertRequest -> upsertSQLStatement(upsertRequest);
+      case RemoveRequest removeRequest -> removeSQLStatement(removeRequest);
+      default ->
+          throw new IllegalStateException(
+              "Request type not implemented for VectorSearch: " + request);
+    };
+  }
+
+  /*
+  Nearest neighbor search helper.
+  */
+  static Result<Stream<NNSearchSQLResultRow>, Exception> StreamSearchResultSet(
+      ResultSet resultSet) {
+    try {
+      return Result.success(
+          JDBCHelper.streamResultSet(
+              resultSet,
+              rs -> {
+                int internalId = rs.getInt("internalId");
+                String queryVectorId = rs.getString("queryVectorId");
+                String neighborId = rs.getString("neighborId");
+                PGvector neighborEmbedding = (PGvector) resultSet.getObject("neighborEmbedding");
+                Double distance = rs.getDouble("distance");
+                return new NNSearchSQLResultRow(
+                    internalId, queryVectorId, neighborId, neighborEmbedding, distance);
+              }));
+    } catch (SQLException e) {
+      return Result.failure(e);
+    }
+  }
+
+  static void addResultRowToMap(
+      Map<Integer, Vectors.Neighbors> integerNeighborsMap,
+      NNSearchSQLResultRow nnSearchSQLResultRow) {
+    Vectors.Datapoint datapoint =
+        new Vectors.Datapoint(
+            nnSearchSQLResultRow.neighborId,
+            JDBCHelper.pGvectorToListDouble(nnSearchSQLResultRow.neighborEmbedding));
+    Vectors.Neighbor neighbor = new Vectors.Neighbor(nnSearchSQLResultRow.distance, datapoint);
+
+    integerNeighborsMap
+        .computeIfAbsent(
+            nnSearchSQLResultRow.internalId,
+            k -> new Vectors.Neighbors(nnSearchSQLResultRow.queryVectorId, new ArrayList<>()))
+        .neighbors()
+        .add(neighbor);
+  }
+
+  static Result<NeighborsResponse, ErrorResponse> neighborsResponseFromResultSet(
+      ResultSet resultSet) {
+    /*
+     * result from DB has <internalId, queryVectorId, neighborId, neighborEmbedding, distance>
+     * this has to be mapped to <<queryVectorId,<neighborId,neighborEmbedding>[]>, distance>[]
+     * respecting the order of the queries in the request
+     */
+    Map<Integer, Vectors.Neighbors> internalIdToNeighborsMap = new LinkedHashMap<>();
+    Result<Stream<NNSearchSQLResultRow>, Exception> nnSearchSQLResultStream =
+        StreamSearchResultSet(resultSet);
+
+    return switch (nnSearchSQLResultStream) {
+      case Result.Failure<?, Exception>(var error) ->
+          Result.failure("Errors occurred while parsing search results", error);
+      case Success<Stream<NNSearchSQLResultRow>, ?>(var resultSetStream) -> {
+        resultSetStream.forEach(
+            nnSearchSQLResultRow -> {
+              addResultRowToMap(internalIdToNeighborsMap, nnSearchSQLResultRow);
```

Comment:
maybe I'm not following, but couldn't this be solved by stream mapping and collecting into a multi map?

Reply:
Updated per your recommendation, please take a look